src/target.rs

```diff
@@ -0,0 +1,180 @@
+//! Very basic parsing of `rustc` target triples.
+//!
+//! See the `target-lexicon` crate for a more principled approach to this.
+
+use std::{borrow::Cow, env, str::FromStr};
+
+use crate::{Error, ErrorKind};
+
+mod generated;
+
+/// The parts of `rustc`'s target triple.
+///
+/// See <https://doc.rust-lang.org/cargo/appendix/glossary.html#target>.
+#[derive(Debug, PartialEq)]
+pub(crate) struct Target {
+    /// The full architecture, including the subarchitecture.
+    ///
+    /// This differs from `cfg!(target_arch)`, which only specifies the
+    /// overall architecture, which is too coarse for certain cases.
+    pub full_arch: Cow<'static, str>,
```

Comment:
I think we can just use `&'a str` here, it'd be simpler and a smaller struct.

For `from_cargo_environment_variables`, we can cache the environment variables in `cc::Build` so that we can return a `Target<'_>`, we already do env caching so this is not a problem.

We can use the [homebrew `OnceLock` impl](https://github.com/rust-lang/cc-rs/blob/main/src/parallel/once_lock.rs) within our codebase, so that we could avoid lifetime problems with `MutexGuard`/`RwLockGuard`.

Reply:
P.S. I'm not worried/obsessed about the allocation as it doesn't matter much, I just prefer a simpler structure.

Caching the env var in `Build` would also be consistent with how other env var currently works:

Once we read the env var from system into `Build`, it's never changed in that object and any objects created by `Build::clone`.

Reply:
Hmm, I don't see how `OnceLock` would help with avoiding a guard? Do you mean to store cached environment variables in a global instead?

Reply:
You can put `OnceLock` inside `cc::Build` struct.

For example:

```rust
struct Build {
    target_triple: OnceLock<String>,
    // ...
}
```

store individual fields of `Target` directly in it, and return `Target<'_>` with a lifetime

Reply:
Hmm, I'm still not _entirely_ sure I get what you want me to do, and I feel like it'd make the code more convoluted?

Would you mind doing it?

Reply:
Sure leave it to me

Reply:
Sure leave it to me

Reply:
I think we could merge it in as-is and then I can open a separate PR to optimize it.